### PR TITLE
The Initial version of CLP in CPU backend

### PR DESCRIPTION
**Pull request by elvinhajizada**
Issue Number: #706 

Objective of pull request:
As a user, I want to be able to use the CLP algorithm in Lava, to learn continually from a stream of data in Loihi chip.

## Pull request checklist

Your PR fulfills the following requirements:
-   [x] [Issue](https://github.com/lava-nc/lava/issues/706) created that explains the change and why it's needed
-   [x] Tests are part of the PR (for bug fixes / features)
-   [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
-   [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
-   [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
-   [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
-   [x] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

-   [ ] Bugfix
-   [x] Feature
-   [ ] Code style update (formatting, renaming)
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Build related changes
-   [ ] Documentation changes
-   [ ] Other (please describe): 

## What is the current behavior?

- We don't have any continual learning algorithm in Lava

## What is the new behavior?

This PR is the first step in the full implementation of CLP. With this version of CLP, three base processes are implemented in the CPU backend and have been tested in terms of the following behaviors:
- PrototypeLIF process generates the correct bAP signal if it receives a 3rd-factor input
- PrototypeLIF process overwrites y1 trace with the value of the received 3rd-factor input, every time such an input is injected
- The novelty can be detected from the coincidence between the presence of input and the absence of a subsequent output
- Readout process outputs the predicted label
- Readout process labels unlabeled winner neurons with user-provided label
- Readout process checks the correctness of the predicted label against the user-provided actual label

In addition, several integration tests are written to test if LearningDense -> PrototypeLIF -> NoveltyDetection intengration works properly

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

**Comment by elvinhajizada** [comment]
All check have passes

**Comment by elvinhajizada** [comment]
Integration tests are added

**Comment by elvinhajizada** [comment]
> Just glanced over it, will review in more detail after some questions.
> 
> 1. Why did you call it prototype_lif? It seems like a 3F learning LIF neuron. Also, is this usable in general? If yes, we should not hide it inside clp folder.
> 2. What does nsm mean? The process is called Readout, I could not figure out what nsm stands for.
> 3. I saw print statements in the PyNoveltyDetectorModel. I assume they are left overs from the development?

@PhilippPlank 
1. Prototype LIF right now looks like a 3F learning LIF, because it is the first, simple of version. In upcoming releases we will add more features to this neuron model that will make it different from LIF (adaptive learning rate, adaptive threshold etc.) So right now it is just setting the interface that will be filled in later releases. 
2.  NSM means "Neural State Machines". I planned it as a folder that will host possibly multiple processes. Right now it has only Readout process, because in this release of CLP, we just reading out network prediction. But later, we will need to receive user labels and create supervisory signals to inject into the network. Based on these we will perform both supervised and unsupervised learning. For that purpose we will need other CPU processes as interface between user and network in Loihi 2. So NSM is a general name for bunch of future processes that will control the learning, prediction, labeling states of the system.
3. Yes, right I will remove those. Thanks for pointing out.

File: tests/lava/proc/clp/nsm/test_nsm.py
[the remark this answers is not in the document]
Comment (elvinhajizada) [reply tests/lava/proc/clp/nsm/test_nsm.py:1]: They should be all resolved now

**Review by PhilippPlank — APPROVED** [review]

**Review by drager-intel — APPROVED** [review]
Looks good now that all interfaces changes that we discussed have been incorporated. Thanks!

File: src/lava/proc/clp/novelty_detector/process.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        super().__init__(t_wait=t_wait, n_protos=n_protos)
+
+        # An input is being processed by the system
+        self.input_aval_in = InPort(shape=(1,))
+
+        # An output is generated by the system
+        self.output_aval_in = InPort(shape=(1,))
+
+        # OutPort for sending out the novelty detection signal
+        self.novelty_detected_out = OutPort(shape=(n_protos,))
```
[the remark this answers is not in the document]
Comment (elvinhajizada) [reply src/lava/proc/clp/novelty_detector/process.py]: he shape of the outport is (1,) now

File: src/lava/proc/clp/novelty_detector/process.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            reset and NO novelty detection signal will be sent out.
+        n_protos: int
+            The number of prototypes that this novelty detector process can
+            target. Each time a novel input is detected the next unallocated
+            prototype will be targeted by the output of the NoveltyDetector
+            process.
+    """
+
+    def __init__(self, *,
+                 t_wait: int,
```
[the remark this answers is not in the document]
Comment (elvinhajizada) [reply src/lava/proc/clp/novelty_detector/process.py]: t_wait is var now and n_protos is removed from novelty detector, as there is no need

File: src/lava/proc/clp/novelty_detector/models.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    novelty_detected_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, np.int32)
+
+    def __init__(self, proc_params):
+        super().__init__(proc_params)
+        self.t_wait = proc_params['t_wait']
+        self.n_protos = proc_params['n_protos']
+        self.waiting = False  # A variable to know if we are waiting for output
+        self.t_passed = 0  # The time passed since the injection of the input
+        self.next_alloc_id = 0  # The id of the next neuron to be allocated
```
[the remark this answers is not in the document]
Comment (elvinhajizada) [reply src/lava/proc/clp/novelty_detector/models.py]: this is moved to Allocator and is also a Var now

File: src/lava/proc/conv/utils.py [hunk 4]
```diff
@@ -482,7 +482,10 @@ def dst_id(x: int, y: int, c: int):
...
 
-    srcs = np.concatenate(srcs).astype(int)
-    dsts = np.concatenate(dsts).astype(int)
-    wgts = np.concatenate(wgts)
+    if len(srcs) >= 1:
+        srcs = np.concatenate(srcs).astype(int)
+    if len(dsts) >= 1:
+        dsts = np.concatenate(dsts).astype(int)
+    if len(wgts) >= 1:
+        wgts = np.concatenate(wgts)
```
Comment (mgkwill) [inline src/lava/proc/conv/utils.py:490]: Fix for test_conv_to_sparse failure. Not my code or unit test. @bamsumit  is this your code/test and if so does this fix look problematic at all?

```
======================================================================
ERROR: test_conv_to_sparse (tests.lava.proc.conv.test_utils.TestConv)
Tests translation of a conv to a sparse layer
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/lava/lava/tests/lava/proc/conv/test_utils.py", line 138, in test_conv_to_sparse
    dst, src, wgt = utils.conv_to_sparse(input_shape=input.shape,
  File "/Users/runner/work/lava/lava/src/lava/proc/conv/utils.py", line 488, in conv_to_sparse
    dsts = np.concatenate(dsts).astype(int)
  File "<__array_function__ internals>", line 200, in concatenate
ValueError: need at least one array to concatenate

----------------------------------------------------------------------
Ran 566 tests in 190.079s

FAILED (errors=1, skipped=18)
Error: Process completed with exit code 1.
```

Comment (elvinhajizada) [reply src/lava/proc/conv/utils.py]: The issue is resolved by Sumit now

**Pull request merged**